working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,612 @@
+
+> Please update this function so that we could use it for any model class
+
+The response delivers exactly what we're looking for:
+```py
+def validate_model(model_class, model_id):
+    try:
+        model_id = int(model_id)
+    except:
+        abort(make_response({"message": f"{model_class.__name__.lower()} {model_id} invalid"}, 400))
+
+    model = model_class.query.get(model_id)
+
+    if not model:
+        abort(make_response({"message": f"{model_class.__name__.lower()} {model_id} not found"}, 404))
+
+    return model
+```
+
+Our function is updated, now we just need to update our dependent functions and tests to use `validate_model` instead of `validate_book`, which includes passing in the new `model_class` parameter. We may also decide to rename `model_class` to the more common `cls`, which is often used in Python when passing in a class reference. We can do all of this manually using VS Code's Find & Replace tools followed by some minor edits to add our new parameter, or we can reach out to Copilot.
+
+To rename the class reference parameter, we select the `validate_model` function again, open the inline Copilot chat, and provide the following prompt:
+
+> Please use cls for the class parameter name
+
+Find & Replace could have done this just as easily, and we might have had more confidence in the results, but changing the rest of the code to use the refactored function is a bit more complex. Let's try having Copilot helping us with the rest of the change. 
+
+One approach we can try is to open each file, select the entire contents, and ask Copilot to change any instances of `validate_book` to `validate_model`. We need to carefully review the lines Copilot suggests changing to ensure nothing unexpected is altered, but it can save us a little time since we need to do slightly more than find and replace the function name.
+
+Due to differences in context, it's possible that a prompt that works fine in one file may not work as well in another file. For example, in the `routes.py` file, the following prompt, which provides very few details, is able to produce the desired changes:
+
+> Please replace all uses of validate_book with validate_model
+
+However, in the `test_routes.py` file, more detail is necessary for Copilot to recognize that there is a new class parameter which is required wherever the function is invoked, and that the `Book` class needs to be imported:
+
+> Please replace all uses of validate_book(book_id) with validate_model(cls, model_id). Import any required model classes.
+
+Since there is always a degree of validation and review required with any change when using Copilot, if we're having trouble finding an effective prompt, it may be worth doing the refactoring some other way. We can always refactor by hand, or by using dedicated refactoring utilities or those built into a development environment. Copilot is just one tool among many that we can draw on as developers.
```

Comment:
Another case where I could have done better with my initial suggestion. Splitting the three alternatives into clearly separate cases reads with less confusion for me.
```suggestion
Since there is always a degree of validation and review required with any change when using Copilot, if we're having trouble finding an effective prompt, it may be worth doing the refactoring some other way. We can always refactor by hand, by using standalone refactoring tools, or by using tools in our development environment. Copilot is just one tool among many that we can draw on as developers.
```